NotificationHubs/src/com/windowsazure/messaging/BaseInstallation.java

```diff
@@ -0,0 +1,310 @@
+     * @param value The platform for the installation.
+     */
+    public void setPlatform(NotificationPlatform value) {
+        // Validate that this is not FCM
+        validateNotificationPlatform(value);
+
+        platform = value;
+    }
+
+    /**
+     * Gets the user ID for the installation.
+     *
+     * @return The user ID for the installation.
+     */
+    public String getUserId() {
+        return userId;
+    }
+
+    /**
+     * Sets the user ID for the installation.
+     *
+     * @param value The user ID for the installation.
+     */
+    public void setUserId(String value) {
+        userId = value;
+    }
+
+    /**
+     * Gets the tags for the installation.
+     *
+     * @return The tags for the installation.
+     */
+    public List<String> getTags() {
+        return tags;
+    }
+
+    /**
+     * Adds a tag to the installation.
+     *
+     * @param tag The tag to add to the installation;
+     */
+    public void addTag(String tag) {
+        if (tags == null) {
+            tags = new ArrayList<>();
+        }
+
+        tags.add(tag);
+    }
+
+    /**
+     * Removes a tag from the installation.
+     *
+     * @param tag The tag to remove from the installation.
+     */
+    public void removeTag(String tag) {
+        if (tags == null) {
+            return;
+        }
+
+        tags.remove(tag);
+    }
+
+    /**
+     * Clears the tags for the installation.
+     */
+    public void clearTags() {
+        if (tags == null) {
+            return;
+        }
+
+        tags.clear();
+    }
+
+    /**
+     * Gets the installation templates for the installation.
+     *
+     * @return The installation templates for the installation.
+     */
+    public Map<String, InstallationTemplate> getTemplates() {
+        return templates;
+    }
+
+    /**
+     * Adds an installation template by name to the installation.
+     *
+     * @param templateName The name for the installation template.
+     * @param template     The template to add to the installation.
+     */
+    public void addTemplate(String templateName, InstallationTemplate template) {
+        if (templates == null) {
+            templates = new HashMap<>();
+        }
+
+        templates.put(templateName, template);
+    }
+
+    /**
+     * Removes an installation template based upon the template name.
+     *
+     * @param templateName The name of the installation template to remove.
+     */
+    public void removeTemplate(String templateName) {
+        if (templates == null) {
+            return;
+        }
+
+        templates.remove(templateName);
+    }
+
+    /**
+     * Clears the installation templates.
+     */
+    public void clearTemplates() {
+        if (templates == null) {
+            return;
+        }
+
+        templates.clear();
+    }
+
+    /**
+     * Gets the secondary tiles for WNS
+     *
+     * @return The secondary tiles for WNS.
+     */
+    public Map<String, WnsSecondaryTile> getSecondaryTiles() {
```

Comment:
We should definitely pull this out of the `BaseInstallation` class and associate it only with `WnsInstallations`.

Reply:
Same for all of the helpers regarding secondary tiles below as well.

Reply:
To do that, we'd need to subclass the WnsInstallation and also add methods for retrieving one just as we did for BrowserInstallation.  Do you want to go down that road?  We certainly can...

Reply:
That's what I've been doing :) I think it's cleaner

Reply:
Done.